src/entities/causalChain.ts

```diff
@@ -0,0 +1,263 @@
+        Link.counter++;
+        this.ingress = new Collections.Set<Link<S, T>>();
+        this.egress = new Collections.Set<Link<S, T>>();
+    }
+
+    /**
+     * The causes.
+     * @returns {Immutable.Set<S>}
+     */
+    public get causes(): Immutable.Set<S> {
+        return this.before;
+    }
+
+    /**
+     * The effects.
+     * @returns {Immutable.Set<S>}
+     */
+    public get effects(): Immutable.Set<S> {
+        return this.after;
+    }
+
+    /**
+     * The causally antecedent links.
+     * @returns {Set<Link<S, T>>}
+     */
+    public get antecedent(): Collections.Set<Link<S, T>> {
+        return this.ingress;
+    }
+
+    /**
+     * The causally incident links.
+     * @returns {Set<Link<S, T>>}
+     */
+    public get incident(): Collections.Set<Link<S, T>> {
+        return this.egress;
+    }
+
+    /**
+     * Retrieve the event represented by this link.
+     * @returns {T}
+     */
+    public getEvent(): T {
+        return this.event;
+    }
+
+    /**
+     * Link this chain to another chain,
+     * which will then exhibit causal incidence.
+     * Returns whether or not the act of linking was successful.
+     * @param {Link<S, T>} other
+     * @returns {boolean}
+     */
+    public linkTo(other: Link<S, T>): boolean {
+        let reachable = false;
+        this.effects.forEach(c => {
+            if (c !== undefined && other.causes.contains(c)) {
+                reachable = true;
+            }
+        });
+
+        if (reachable) {
+            this.incident.add(other);
+            other.antecedent.add(this);
+            return true;
+        }
+
+        return false;
+    }
+
+    /**
+     * Find an ancestor of this link with an event identical to that of a given link.
+     * Returns whether a similar ancestor was found.
+     * @param {Link<S, T>} toMatch
+     * @returns {Link<S, T> | undefined}
+     */
+    public hasSimilarAncestor(toMatch: Link<S, T>): boolean {
+        let matchFound = false;
+        this.antecedent.forEach(ancestor => {
+            if (ancestor.getEvent() === toMatch.getEvent()) {
+                matchFound = true;
+                return;
+            } else {
+                matchFound = ancestor.hasSimilarAncestor(toMatch);
+                if (matchFound) {
+                    return;
+                }
+            }
+        });
+
+        return matchFound;
+    }
+
+    /**
+     * Severs the connection between this link and an ancestor link.
+     * Does nothing when the given link is not an ancestor of the current link.
+     * @param {Link<S, T>} ancestor
+     */
+    public severLink(ancestor: Link<S, T>): void {
+        ancestor.incident.remove(this);
+        this.antecedent.remove(ancestor);
+    }
+
+    /**
+     * Retrieve a string representation of this link.
+     * Used simply to uniquely differentiate from other links.
+     * @returns {string}
+     */
+    public toString(): string {
+        return "" + this.identifier;
```

Comment:
Just call `this.identifier.toString()` here please. There's really no reason to golf this.

Reply:
I tried that; didn't seem to work. Maybe I did something wrong so I'll verify.

Reply:
JavaScript is kind of "funny:" `1.toString()` is a syntax error, but `let x = 1; x.toString()` is perfectly legitimate.